File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/Grammar.java
package titan.ast.runtime;

/**
 * 语法（文法）.
 *
 * @author tian wei jun
 */
public class Grammar {
  public int index;
  public String name = "";
  public GrammarType type = GrammarType.TERMINAL;
  public GrammarAction action = GrammarAction.TEXT;

  public Grammar(int index) {
    this.index = index;
  }

  @Override
  public String toString() {
    return name;
  }

  public AstGrammar toAstGrammar() {
    return new AstGrammar(name, type);
  }
}

File: runtime/cpp/titan-ast-runtime-lib/Grammar.cpp
//
// Created by tian wei jun on 2022/11/17 0017.
//

#include "Grammar.h"

Grammar::Grammar(int index)
    : index(index), name(std::string()), type(GrammarType::TERMINAL),
      action(GrammarAction::TEXT) {}

Grammar &Grammar::operator=(const Grammar &other) = default;

Grammar::~Grammar() = default;

// ------------------TerminalGrammar----------------------

TerminalGrammar::TerminalGrammar(int index)
    : Grammar(index), lookaheadMatchingMode(LookaheadMatchingMode::GREEDINESS) {
  this->type = GrammarType::TERMINAL;
}

TerminalGrammar::~TerminalGrammar() = default;

// ------------------NonterminalGrammar----------------------
NonterminalGrammar::NonterminalGrammar(int index) : Grammar(index) {
  this->type = GrammarType::NONTERMINAL;
}

NonterminalGrammar::~NonterminalGrammar() = default;